### PR TITLE
US-5: Start building for travel plan

### PR DESCRIPTION
**Pull request by leojrsilao**
This PR implements #75.

When a user taps on the start/end location input and taps on a building, the building name will be filled in. (Completed in Task-22 #74 )

When locations services are on:
- If the user is in a campus building, the start location input will be filled in. 
- The user is able to search for the current location as a starting location and it will always be the first result in the search suggestions.

### How to test if the user is in a campus building (Hall building)
Add the following code in the file or run the program in an emulator and mock the location:
```
map-screen.component.tsx:
if (isPointInPolygon({ latitude: 45.497237, longitude: -73.578913 }, 
building.boundingBox)) 
{          onBuildingTap(building);
           setStartLocationDisplay(building.displayName);
}
```

NOTE:
Acceptance tests will be done in #92.

### CODE REVIEW

**Review by RGPosadas — CHANGES_REQUESTED** [review]
#### BLOCKER 1: Unit Tests
I don't see a unit test that tests for `Current Location` to be the startLocation. Create one as you see fit.

File: __tests__/utility.service.test.js [hunk 1]
```diff
@@ -19,18 +31,21 @@ describe("UtilityService", () => {
   });
 
   describe("updateSearchResults", () => {
-    it("should call provided functions", () => {
+    it("should call provided functions and ", () => {
```
Comment (RGPosadas) [inline __tests__/utility.service.test.js]: I would prefer to have a more descriptive test name :P

Comment (TheGreatMarkus) [reply __tests__/utility.service.test.js:34]: Yeah its currently incomplete

File: __tests__/utility.service.test.js [hunk 2]
```diff
@@ -19,18 +31,21 @@ describe("UtilityService", () => {
...
         mockSetSearchResults,
-        mockSetDisplayValue
+        mockSetDisplayValue,
+        mockCurrentLocation
       );
 
       expect(mockSetSearchResults).toHaveBeenCalledTimes(1);
       expect(mockSetDisplayValue).toHaveBeenCalledTimes(1);
+      expect(mockCurrentLocation).not.toBe(null);
+      expect(mockCurrentMarkerLocation).not.toBe(null);
```
Comment (cristian-aldea) [inline __tests__/utility.service.test.js]: These tests are a bit useless since you're the one controlling these inputs.

File: src/components/search/omnibox-directions.component.tsx [hunk 3]
```diff
@@ -122,7 +123,7 @@ const OmniboxDirections = ({
     if (currentLocation && !startLocation) {
       setStartLocation({
         id: CURRENT_LOCATION_DISPLAY_TEXT,
-        displayName: CURRENT_LOCATION_DISPLAY_TEXT,
+        displayName: startLocationDisplay,
```
Comment (cristian-aldea) [inline src/components/search/omnibox-directions.component.tsx]: Why was this changed? I feel like it might break something, especially with the new changes merged recently by  #101 and #102. Lines 104 to 120 already take care of syncing up the display text whenever either the start or end locations are updated.

File: src/constants/poi.data.ts [hunk 4]
```diff
@@ -30,6 +30,18 @@ export const POIToSearchResult = (poi: POI): SearchResult => {
...
+): SearchResult => {
+  return {
+    id: markerLocation.id,
+    displayName: markerLocation.displayName,
+    location: markerLocation.location,
+    searchName: markerLocation.displayName,
+    extraInformation: "",
+  };
+};
+
```
Comment (cristian-aldea) [inline src/constants/poi.data.ts]: Please remove this as it was already moved to UtilityService

File: src/services/utility.service.ts [hunk 5]
```diff
@@ -1,10 +1,35 @@
...
+ */
+const MarkerToSearchResult = (markerLocation: MarkerLocation): SearchResult => {
+  return {
+    id: markerLocation.id,
+    displayName: markerLocation.displayName,
+    location: markerLocation.location,
+    searchName: markerLocation.displayName,
+    extraInformation: "",
+  };
+};
```
Comment (cristian-aldea) [inline src/services/utility.service.ts]: Please move this to be inside the class, as a private helper function

File: package-lock.json [hunk 6]
```diff
@@ -1378,6 +1378,12 @@
           "resolved": "https://registry.npmjs.org/async/-/async-0.9.2.tgz",
```
Comment (cristian-aldea) [inline package-lock.json]: Why was the package-lock.json regenerated? I don't see any changes to the package.json.

File: src/components/search/omnibox-directions.component.tsx (hunk 3 above)
Comment (cristian-aldea) [reply src/components/search/omnibox-directions.component.tsx]: Nevermind I see why this was done. But it shouldn't be done like this. 

Please revert this and follow the comment tagged FIX-NOTE-1

File: src/screens/map-screen/map-screen.component.tsx [hunk 7]
```diff
@@ -144,6 +143,7 @@ const MapScreen = () => {
         Buildings.forEach((building) => {
           if (isPointInPolygon(response.coords, building.boundingBox)) {
             onBuildingTap(building);
+            setStartLocationDisplay(building.displayName);
```
Comment (cristian-aldea) [inline src/screens/map-screen/map-screen.component.tsx]: FIX-NOTE-1 (Please refer to the comment above mentionning this one)

Please don't set just the display name, set the actual startLocation as such: 
`setStartLocation(buildingToMarker(building));`

Don't forget to rebase too before doing these changes.

File: src/services/utility.service.ts [hunk 8]
```diff
@@ -64,17 +89,25 @@ class UtilityService {
...
 
+    if (currentLocation) {
+      const currentMarkerLocation: MarkerLocation = {
+        id: CURRENT_LOCATION_DISPLAY_TEXT,
+        displayName: CURRENT_LOCATION_DISPLAY_TEXT,
+        location: currentLocation,
+      };
+      const SearchResultMarker = MarkerToSearchResult(currentMarkerLocation);
+      locations.unshift(SearchResultMarker);
+    }
```
Comment (cristian-aldea) [inline src/services/utility.service.ts]: Actually to make this simpler, you can change MarkerToSearchResult to take in a Location and return a SearchResult, which would save you from having to convert the currentLocation to a MarkerLocation.

Also please rename MarkerToSearchResult to something that doesn't start with an uppercase letter

**Review by cristian-aldea — CHANGES_REQUESTED** [review]
I've left a few comments I'd like to see addressed before we merge.

File: src/components/search/omnibox-directions.component.tsx (hunk 3 above)
Comment (cristian-aldea) [reply src/components/search/omnibox-directions.component.tsx]: Here is why this would be an issue:
image

Notice that the start location isn't set quite correctly (location and id are not from the building object):
image

Instead, these values are set in your changes in omnibox-directions.component.tsx on line 126.

Also, you remarked that you were getting issues with the latest merge #102   with "Fetching Current Location ... " staying and not changing. I'm afraid that issue is probably caused by your changes.

File: package-lock.json (hunk 6 above)
Comment (leojrsilao) [reply package-lock.json]: It was re-generated when I ran npm install

File: src/services/utility.service.ts (hunk 5 above)
Comment (leojrsilao) [reply src/services/utility.service.ts]: Made it into a public helper function and it's used in 2 files.

File: src/constants/poi.data.ts (hunk 4 above)
Comment (leojrsilao) [reply src/constants/poi.data.ts]: Done.

File: src/components/search/omnibox-directions.component.tsx (hunk 3 above)
Comment (leojrsilao) [reply src/components/search/omnibox-directions.component.tsx]: Issue is addressed.

File: __tests__/utility.service.test.js (hunk 2 above)
Comment (leojrsilao) [reply __tests__/utility.service.test.js]: Removed useless tests.

File: package-lock.json (hunk 6 above)
Comment (cristian-aldea) [reply package-lock.json]: You might have regenerated it if you deleted it and ran npm i. It shouldn't be changing it just by running npm i.

Its no big deal but please don't delete package-lock.json unnecessarily.

File: src/services/utility.service.ts [hunk 9]
```diff
@@ -1,12 +1,41 @@
...
+  ): SearchResult => {
+    return {
+      id,
+      displayName,
+      location,
+      searchName,
+      extraInformation,
+    };
+  };
+
```
Comment (cristian-aldea) [inline src/services/utility.service.ts]: Really minor but please put it underneath the fields and the constructor, just like a normal java class :)

File: src/services/utility.service.ts [hunk 10]
```diff
@@ -57,24 +86,33 @@ class UtilityService {
...
-    const locations: SearchResult[] = this.locationsToSearch.filter(
-      (location) => {
-        return (
-          location.searchName.toLowerCase().search(inputText.toLowerCase()) !==
-          -1
-        );
-      }
-    );
+    let locations: SearchResult[] = [];
+    locations = this.locationsToSearch.filter((location) => {
```
Comment (cristian-aldea) [inline src/services/utility.service.ts]: This extra assignment is not necessary, you can just leave it as it was before with `const locations: SearchResult[] = this.locationsToSearch.filter` and then unshift the current location no?

**Review by cristian-aldea — CHANGES_REQUESTED** [review]
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [x] All acceptance tests pass
- [x] If ticket is a user story, there must exist an automated acceptance test for it (NOTE: This US doesn't have one. As they have mentioned, it will be done in #92)

### ACCEPTANCE CRITERIA
- [x] Your Location should pop up as a searchable entry if location services are on

Start building can be filled in by:
- [x] tapping on building polygon/marker (in omnibox focus)
- [x] searching for a building (in omnibox focus)
- [x] location services by automatically filling it in if user is in a building (in omnibox focus)

### USER FLOW ACCEPTANCE TEST
#### Pre-requisite flow
- [x] User taps on search box
- [x] User searches for desired destination first (brings up the omnibox, the points below are all within the omnibox)

#### Main flow
- [x] If location services are on and user is in a building, start location is automatically filled in (the building that the user is in)
- [x] If location services are on and user is NOT in a building, start location is automatically filled in with their current location
- [x] If location services are on and user is not in a building, they can override their start location by tapping on start location box and it can be filled in by tapping on a building or searching for it
- [x] If location services are not on, start location is not automatically filled in. User must set start location (tap or search)


### BLOCKERS

I failed to tell you guys about one change you needed to do for this ticket. As you can see by the unchecked checkboxes and the below screenshot, if I am in a building and location services are used. my start location is set to be my "Current Location" and not the building I am in.

image


Steps to reproduce:
1. start application
2. If using an emulator, set your current location to be in the hall building. If not, change the `getCurrentLocationAsync` method to return:
```js
{
"latitude": 45.49727,
 "longitude": -73.5789317,
}
```
3. Search for a POI or building in the search bar.

This can be fixed by modifying the `setUserCurrentLocation` to include the functionality of `onLocationButtonPress`, and removing `onLocationButtonPress`. I'll be discussing with you to give my suggestion

File: src/services/utility.service.ts (hunk 10 above)
Comment (r-saba) [reply src/services/utility.service.ts:101]: This was at my request. It removes the need for the ternary operator below which should have been removed in the PR but I forgot to mention it. This fixes a bug where the location array wasn't being reset to empty every time the function was called so it would always have 5 values in it even if the text only matched 3.

File: src/services/utility.service.ts (hunk 9 above)
Comment (JimLau49) [reply src/services/utility.service.ts]: sure :)

Comment (JimLau49) [reply src/services/utility.service.ts]: fixed

File: src/services/utility.service.ts (hunk 10 above)
Comment (JimLau49) [reply src/services/utility.service.ts:101]: good catch, fixed

**Review by cristian-aldea — DISMISSED** [review]
User flow acceptance tests pass for me. Good job guys!

File: __tests__/utility.service.test.js [hunk 11]
```diff
@@ -19,18 +20,51 @@ describe("UtilityService", () => {
   });
 
   describe("updateSearchResults", () => {
-    it("should call provided functions", () => {
+    it("should call provided functions and return expected objects", () => {
```
Comment (RGPosadas) [inline __tests__/utility.service.test.js]: Non-blocking: can this be more descriptive? This test name is applicable to pretty much all the unit tests

File: __tests__/utility.service.test.js [hunk 12]
```diff
@@ -19,18 +20,51 @@ describe("UtilityService", () => {
...
+      );
+
+      expect(mockSetSearchResults).toHaveBeenCalledTimes(1);
+      expect(mockSetDisplayValue).toHaveBeenCalledTimes(1);
+      expect(mockSetSearchResults.mock.calls[0][0][0]).toEqual(
+        mockCurrentMarkerLocation
+      );
+    });
+
+    it("Search results and display value should be called once and the mocked value of search results should be empty", () => {
```
Comment (RGPosadas) [inline __tests__/utility.service.test.js]: Non-blocking: nice description, however can this be reworded to match the`it` expectation? Something like `it("should show xxx so that yyy")`

**Review by RGPosadas — DISMISSED** [review]
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [x] All acceptance tests pass
- [x] If ticket is a user story, there must exist an automated acceptance test for it (NOTE: This US doesn't have one. As they have mentioned, it will be done in #92)

### ACCEPTANCE CRITERIA
- [x] Your Location should pop up as a searchable entry if location services are on

Start building can be filled in by:
- [x] tapping on building polygon/marker (in omnibox focus)
- [x] searching for a building (in omnibox focus)
- [x] location services by automatically filling it in if user is in a building (in omnibox focus)

### USER FLOW ACCEPTANCE TEST
#### Pre-requisite flow
- [x] User taps on search box
- [x] User searches for desired destination first (brings up the omnibox, the points below are all within the omnibox)

#### Main flow
- [x] If location services are on and user is in a building, start location is automatically filled in (the building that the user is in)
- [x] If location services are on and user is NOT in a building, start location is automatically filled in with their current location
- [x] If user is not in a building, they can override their current location by tapping on start location box and it can be filled in by tapping on a building or searching for it
- [x] If location services are not on, start location is not automatically filled in. User must set start location (tap or search)

### Notes
Good job guys! I added some non-blocking comments :)

**Review by RGPosadas — APPROVED** [review]
LGTM
